pkg/tarlog/tarlogger.go

```diff
@@ -1,47 +1,65 @@
 package tarlog
 
 import (
-	"archive/tar"
 	"io"
-	"os"
 	"sync"
 
-	"github.com/pkg/errors"
+	"github.com/vbatts/tar-split/archive/tar"
 )
 
 type tarLogger struct {
-	writer *os.File
-	wg     sync.WaitGroup
+	writer     *io.PipeWriter
+	closeMutex *sync.Mutex
+	stateMutex *sync.Mutex
+	closed     bool
 }
 
 // NewLogger returns a writer that, when a tar archive is written to it, calls
 // `logger` for each file header it encounters in the archive.
 func NewLogger(logger func(*tar.Header)) (io.WriteCloser, error) {
-	reader, writer, err := os.Pipe()
-	if err != nil {
-		return nil, errors.Wrapf(err, "error creating pipe for tar logger")
+	reader, writer := io.Pipe()
+	t := &tarLogger{
+		writer:     writer,
+		closeMutex: new(sync.Mutex),
+		stateMutex: new(sync.Mutex),
+		closed:     false,
 	}
-	t := &tarLogger{writer: writer}
 	tr := tar.NewReader(reader)
-	t.wg.Add(1)
+	tr.RawAccounting = true
+	t.closeMutex.Lock()
 	go func() {
 		hdr, err := tr.Next()
 		for err == nil {
 			logger(hdr)
 			hdr, err = tr.Next()
+
 		}
+		// Make sure to avoid writes after the reader has been closed.
+		t.stateMutex.Lock()
+		t.closed = true
 		reader.Close()
-		t.wg.Done()
+		t.stateMutex.Unlock()
+
+		t.closeMutex.Unlock()
 	}()
 	return t, nil
 }
 
 func (t *tarLogger) Write(b []byte) (int, error) {
+	t.stateMutex.Lock()
+	if t.closed {
+		// We cannot use os.Pipe() as this alters the tar's digest. Using
+		// io.Pipe() requires this workaround as it does not allow for writes
+		// after close.
+		t.stateMutex.Unlock()
+		return len(b), nil
+	}
+	t.stateMutex.Unlock()
 	return t.writer.Write(b)
 }
 
 func (t *tarLogger) Close() error {
-	err := t.writer.Close()
-	t.wg.Wait()
-	return err
+	t.writer.Close()
+	t.closeMutex.Lock()
```

Comment:
OK, maybe late Saturday brain, but should this be Unlock?

Reply:
That's intended. I changed the waiting group to a mutex as it was used to always sync two routines. The lock makes sure to wait for the other routine.